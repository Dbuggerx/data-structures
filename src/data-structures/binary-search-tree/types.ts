export type NodeKey = number | string | ICompositeKey;

export interface ICompositeKey {
  compareTo: (other: any) => Comparison;
}

export enum Comparison {
  Smaller,
  Equal,
  Greater
}
