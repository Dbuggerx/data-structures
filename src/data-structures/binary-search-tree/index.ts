import TreeNode from "./tree-node";
import { NodeKey, Comparison } from "./types";

export default class BinarySearchTree<T> {
  protected rootNode: TreeNode<T> | undefined;

  get root() {
    return this.rootNode as Readonly<TreeNode<T> | undefined>;
  }

  private findNode(key: NodeKey, node?: TreeNode<T>): TreeNode<T> | undefined {
    if (!node || node.compareToKey(key) === Comparison.Equal) return node;
    return node.compareToKey(key) === Comparison.Greater
      ? this.findNode(key, node.left)
      : this.findNode(key, node.right);
  }

  protected getNodeHeight(node: TreeNode<T>): number {
    return Math.max(node.left?.height || 0, node.right?.height || 0) + 1;
  }

  protected insertNode(nodeToAdd: TreeNode<T>, currentNode?: TreeNode<T>) {
    if (!currentNode) return nodeToAdd;

    if (currentNode.compareToKey(nodeToAdd.key) === Comparison.Greater)
      currentNode.left = this.insertNode(nodeToAdd, currentNode.left);
    else currentNode.right = this.insertNode(nodeToAdd, currentNode.right);

    currentNode.height = this.getNodeHeight(currentNode);

    return currentNode;
  }

  protected removeNode(key: NodeKey, currentNode?: TreeNode<T>) {
    if (!currentNode) return undefined;

    const currentNodeComparison = currentNode.compareToKey(key);

    // If the key to be deleted is smaller than the node's key, then it lies in the left subtree
    if (currentNodeComparison === Comparison.Greater)
      currentNode.left = this.removeNode(key, currentNode.left);
    // If the key to be deleted is greater than the node's key, then it lies in the right subtree
    else if (currentNodeComparison === Comparison.Smaller)
      currentNode.right = this.removeNode(key, currentNode.right);
    // if key is same as node's key, then this is the node to be deleted
    else if (!currentNode.left || !currentNode.right) {
      return currentNode.left || currentNode.right;
    } else {
      // node with two children: get the inorder successor (smallest in the right subtree)
      const temp = this.getMinValueNode(currentNode.right);

      // Copy the inorder successor's data to this node
      currentNode.key = temp.key;
      currentNode.value = temp.value;

      // Delete the inorder successor
      currentNode.right = this.removeNode(temp.key, currentNode.right);
    }

    currentNode.height = this.getNodeHeight(currentNode);
    return currentNode;
  }

  private getMinValueNode(node: TreeNode<T>) {
    let current = node;

    // loop down to find the leftmost leaf
    while (current.left) current = current.left;

    return current;
  }

  add(key: NodeKey, value: T) {
    const nodeToAdd = new TreeNode(key, value);
    this.rootNode = this.insertNode(nodeToAdd, this.rootNode);
    return nodeToAdd as Readonly<TreeNode<T>>;
  }

  remove(key: NodeKey) {
    this.rootNode = this.removeNode(key, this.rootNode);
    return this;
  }

  find(key: NodeKey): T | undefined {
    return this.findNode(key, this.rootNode)?.value;
  }

  *inOrderTraversal(
    node: TreeNode<T> | undefined = this.rootNode
  ): IterableIterator<Readonly<NodeEntry<T>>> {
    if (!node) return;

    if (node.left) yield* this.inOrderTraversal(node.left);
    yield { key: node.key, value: node.value };
    if (node.right) yield* this.inOrderTraversal(node.right);
  }

  *reverseInOrderTraversal(
    node: TreeNode<T> | undefined = this.rootNode
  ): IterableIterator<Readonly<NodeEntry<T>>> {
    if (!node) return;

    if (node.right) yield* this.reverseInOrderTraversal(node.right);
    yield { key: node.key, value: node.value };
    if (node.left) yield* this.reverseInOrderTraversal(node.left);
  }
}

type NodeEntry<T> = Pick<TreeNode<T>, "key" | "value">;
