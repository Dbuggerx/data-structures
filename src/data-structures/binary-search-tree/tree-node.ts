import { NodeKey, Comparison } from "./types";

export default class TreeNode<T> {
  key: NodeKey;
  value: T;
  height: number;
  left: TreeNode<T> | undefined;
  right: TreeNode<T> | undefined;

  constructor(key: NodeKey, value: T) {
    this.key = key;
    this.value = value;
    this.height = 1;
  }

  compareToKey(otherKey: NodeKey): Comparison {
    if (typeof this.key === "number" || typeof this.key === "string") {
      if (this.key < otherKey) return Comparison.Smaller;
      else if (this.key > otherKey) return Comparison.Greater;
      else return Comparison.Equal;
    }

    return this.key.compareTo(otherKey);
  }
}
