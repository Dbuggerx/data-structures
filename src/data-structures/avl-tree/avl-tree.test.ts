import AvlTree from "./index";
import { ICompositeKey, Comparison } from "../binary-search-tree/types";

describe("AVL Tree", () => {
  describe("primitive keys", () => {
    describe("add", () => {
      it("adds root node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(10, "a");
        avlTree.add(20, "b");

        expect(avlTree.root).toEqual(
          expect.objectContaining({
            key: 10,
            value: "a",
            height: 2
          })
        );
      });

      it("returns added node", () => {
        const avlTree = new AvlTree<string>();

        expect(avlTree.add(10, "a")).toEqual({
          key: 10,
          value: "a",
          height: 1
        });

        expect(avlTree.add(20, "b")).toEqual({
          key: 20,
          value: "b",
          height: 1
        });
      });

      it("adds nested nodes", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(10, "a");
        avlTree.add(9, "b");
        avlTree.add(8, "c");
        avlTree.add(90, "d");
        avlTree.add(190, "e");
        avlTree.add(5, "f");
        avlTree.add(6, "g");
        avlTree.add(2, "h");
        avlTree.add(80, "i");
        avlTree.add(3, "j");
        avlTree.add(70, "k");
        avlTree.add(4, "l");

        expect(avlTree.root).toMatchSnapshot();

        avlTree.add(7, "m");
        expect(avlTree.root).toMatchSnapshot();
      });
    });

    describe("traversal", () => {
      let avlTree: AvlTree<string>;

      beforeEach(() => {
        avlTree = new AvlTree();
        avlTree.add(20, "a");
        avlTree.add(10, "b");
        avlTree.add(30, "c");
        avlTree.add(15, "d");
        avlTree.add(50, "e");
      });

      it("iterates in order", () => {
        expect(Array.from(avlTree.inOrderTraversal())).toEqual([
          { key: 10, value: "b" },
          { key: 15, value: "d" },
          { key: 20, value: "a" },
          { key: 30, value: "c" },
          { key: 50, value: "e" }
        ]);
      });

      it("iterates in reverse order", () => {
        expect(Array.from(avlTree.reverseInOrderTraversal())).toEqual([
          { key: 50, value: "e" },
          { key: 30, value: "c" },
          { key: 20, value: "a" },
          { key: 15, value: "d" },
          { key: 10, value: "b" }
        ]);
      });
    });

    describe("remove", () => {
      it("removes root node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(10, "a");
        expect(avlTree.root?.key).toBe(10);

        avlTree.remove(10);
        expect(avlTree.root).toBeUndefined();
      });

      it("does nothing for non existing node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(10, "a");

        avlTree.remove(1);
        expect(avlTree.root?.key).toBe(10);
      });

      it("removes nested nodes", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(10, "a");
        avlTree.add(9, "b");
        avlTree.add(8, "c");
        avlTree.add(90, "d");
        avlTree.add(190, "e");
        avlTree.add(5, "f");
        avlTree.add(6, "g");
        avlTree.add(2, "h");
        avlTree.add(80, "i");
        avlTree.add(3, "j");
        avlTree.add(70, "k");
        avlTree.add(4, "l");

        avlTree.remove(9);
        expect(avlTree.root).toMatchSnapshot();

        avlTree.remove(80);
        expect(avlTree.root).toMatchSnapshot();
      });
    });

    describe("find", () => {
      it("finds existing nodes", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(20, "a");
        avlTree.add(10, "b");
        avlTree.add(30, "c");

        expect(avlTree.find(30)).toEqual("c");
        expect(avlTree.find(10)).toEqual("b");
        expect(avlTree.find(20)).toEqual("a");
      });

      it("does not find non existing node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(20, "a");
        avlTree.add(10, "b");
        avlTree.add(30, "c");

        expect(avlTree.find(50)).toBeUndefined();
      });
    });
  });

  describe("composite keys", () => {
    class DummyCompositeKey implements ICompositeKey {
      constructor(private id: number, private name: string) {}

      compareTo(other: DummyCompositeKey) {
        return this.id === other.id ? this.compareToName(other) : this.compareToId(other);
      }

      compareToId(other: DummyCompositeKey) {
        if (this.id > other.id) return Comparison.Greater;
        else if (this.id < other.id) return Comparison.Smaller;
        return Comparison.Equal;
      }

      compareToName(other: DummyCompositeKey) {
        if (this.name > other.name) return Comparison.Greater;
        else if (this.name < other.name) return Comparison.Smaller;
        return Comparison.Equal;
      }
    }

    describe("add", () => {
      it("adds root node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");
        avlTree.add(new DummyCompositeKey(20, "bbb"), "b");

        expect(avlTree.root).toEqual(
          expect.objectContaining({
            key: {
              id: 10,
              name: "aaa"
            },
            value: "a",
            height: 2
          })
        );
      });

      it("returns added node", () => {
        const avlTree = new AvlTree<string>();

        expect(avlTree.add(new DummyCompositeKey(10, "aaa"), "a")).toEqual({
          key: {
            id: 10,
            name: "aaa"
          },
          value: "a",
          height: 1
        });

        expect(avlTree.add(new DummyCompositeKey(10, "bbb"), "b")).toEqual({
          key: {
            id: 10,
            name: "bbb"
          },
          value: "b",
          height: 1
        });

        expect(avlTree.add(new DummyCompositeKey(20, "ccc"), "c")).toEqual({
          key: {
            id: 20,
            name: "ccc"
          },
          value: "c",
          height: 1
        });
      });

      it("adds nested nodes", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");
        avlTree.add(new DummyCompositeKey(9, "bbb"), "b");
        avlTree.add(new DummyCompositeKey(8, "ccc"), "c");
        avlTree.add(new DummyCompositeKey(10, "ddd"), "d");
        avlTree.add(new DummyCompositeKey(20, "eee"), "e");
        avlTree.add(new DummyCompositeKey(9, "fff"), "f");

        expect(avlTree.root).toMatchSnapshot();
      });
    });

    describe("traversal", () => {
      let avlTree: AvlTree<string>;

      beforeEach(() => {
        avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");
        avlTree.add(new DummyCompositeKey(9, "bbb"), "b");
        avlTree.add(new DummyCompositeKey(8, "ccc"), "c");
        avlTree.add(new DummyCompositeKey(10, "ddd"), "d");
        avlTree.add(new DummyCompositeKey(20, "eee"), "e");
        avlTree.add(new DummyCompositeKey(9, "fff"), "f");
      });

      it("iterates in order", () => {
        expect(Array.from(avlTree.inOrderTraversal())).toEqual([
          { key: { id: 8, name: "ccc" }, value: "c" },
          { key: { id: 9, name: "bbb" }, value: "b" },
          { key: { id: 9, name: "fff" }, value: "f" },
          { key: { id: 10, name: "aaa" }, value: "a" },
          { key: { id: 10, name: "ddd" }, value: "d" },
          { key: { id: 20, name: "eee" }, value: "e" }
        ]);
      });

      it("iterates in reverse order", () => {
        expect(Array.from(avlTree.reverseInOrderTraversal())).toEqual([
          { key: { id: 20, name: "eee" }, value: "e" },
          { key: { id: 10, name: "ddd" }, value: "d" },
          { key: { id: 10, name: "aaa" }, value: "a" },
          { key: { id: 9, name: "fff" }, value: "f" },
          { key: { id: 9, name: "bbb" }, value: "b" },
          { key: { id: 8, name: "ccc" }, value: "c" }
        ]);
      });
    });

    describe("remove", () => {
      it("removes root node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");

        avlTree.remove(new DummyCompositeKey(10, "aaa"));
        expect(avlTree.root).toBeUndefined();
      });

      it("does nothing for non existing node", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");

        avlTree.remove(new DummyCompositeKey(10, "bbb"));
        expect(avlTree.root?.key).toEqual({
          id: 10,
          name: "aaa"
        });
      });

      it("removes nested nodes", () => {
        const avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(10, "aaa"), "a");
        avlTree.add(new DummyCompositeKey(9, "bbb"), "b");
        avlTree.add(new DummyCompositeKey(8, "ccc"), "c");
        avlTree.add(new DummyCompositeKey(90, "ddd"), "d");
        avlTree.add(new DummyCompositeKey(190, "eee"), "e");
        avlTree.add(new DummyCompositeKey(5, "fff"), "f");
        avlTree.add(new DummyCompositeKey(6, "ggg"), "g");
        avlTree.add(new DummyCompositeKey(2, "hhh"), "h");
        avlTree.add(new DummyCompositeKey(80, "iii"), "i");
        avlTree.add(new DummyCompositeKey(3, "jjj"), "j");
        avlTree.add(new DummyCompositeKey(70, "kkk"), "k");
        avlTree.add(new DummyCompositeKey(4, "lll"), "l");
        avlTree.add(new DummyCompositeKey(70, "mmm"), "m");

        avlTree.remove(new DummyCompositeKey(70, "kkk"));
        expect(avlTree.root).toMatchSnapshot();

        avlTree.remove(new DummyCompositeKey(8, "ccc"));
        expect(avlTree.root).toMatchSnapshot();
      });
    });

    describe("find", () => {
      let avlTree: AvlTree<string>;

      beforeEach(() => {
        avlTree = new AvlTree<string>();
        avlTree.add(new DummyCompositeKey(20, "aaa"), "a");
        avlTree.add(new DummyCompositeKey(10, "bbb"), "b");
        avlTree.add(new DummyCompositeKey(30, "ccc"), "c");
        avlTree.add(new DummyCompositeKey(10, "ddd"), "d");
      });

      it("finds existing nodes", () => {
        expect(avlTree.find(new DummyCompositeKey(30, "ccc"))).toEqual("c");
        expect(avlTree.find(new DummyCompositeKey(10, "bbb"))).toEqual("b");
        expect(avlTree.find(new DummyCompositeKey(20, "aaa"))).toEqual("a");
      });

      it("does not find non existing node", () => {
        expect(avlTree.find(new DummyCompositeKey(50, "aaa"))).toBeUndefined();
      });
    });
  });
});
