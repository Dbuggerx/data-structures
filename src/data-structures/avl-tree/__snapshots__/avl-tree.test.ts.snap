// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AVL Tree composite keys add adds nested nodes 1`] = `
TreeNode {
  "height": 3,
  "key": DummyCompositeKey {
    "id": 10,
    "name": "aaa",
  },
  "left": TreeNode {
    "height": 2,
    "key": DummyCompositeKey {
      "id": 9,
      "name": "bbb",
    },
    "left": TreeNode {
      "height": 1,
      "key": DummyCompositeKey {
        "id": 8,
        "name": "ccc",
      },
      "value": "c",
    },
    "right": TreeNode {
      "height": 1,
      "key": DummyCompositeKey {
        "id": 9,
        "name": "fff",
      },
      "value": "f",
    },
    "value": "b",
  },
  "right": TreeNode {
    "height": 2,
    "key": DummyCompositeKey {
      "id": 10,
      "name": "ddd",
    },
    "left": undefined,
    "right": TreeNode {
      "height": 1,
      "key": DummyCompositeKey {
        "id": 20,
        "name": "eee",
      },
      "value": "e",
    },
    "value": "d",
  },
  "value": "a",
}
`;

exports[`AVL Tree composite keys remove removes nested nodes 1`] = `
TreeNode {
  "height": 4,
  "key": DummyCompositeKey {
    "id": 9,
    "name": "bbb",
  },
  "left": TreeNode {
    "height": 3,
    "key": DummyCompositeKey {
      "id": 5,
      "name": "fff",
    },
    "left": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 3,
        "name": "jjj",
      },
      "left": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 2,
          "name": "hhh",
        },
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 4,
          "name": "lll",
        },
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 6,
        "name": "ggg",
      },
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 8,
          "name": "ccc",
        },
        "left": undefined,
        "value": "c",
      },
      "value": "g",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 3,
    "key": DummyCompositeKey {
      "id": 80,
      "name": "iii",
    },
    "left": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 70,
        "name": "mmm",
      },
      "left": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 10,
          "name": "aaa",
        },
        "left": undefined,
        "right": undefined,
        "value": "a",
      },
      "right": undefined,
      "value": "m",
    },
    "right": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 90,
        "name": "ddd",
      },
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 190,
          "name": "eee",
        },
        "value": "e",
      },
      "value": "d",
    },
    "value": "i",
  },
  "value": "b",
}
`;

exports[`AVL Tree composite keys remove removes nested nodes 2`] = `
TreeNode {
  "height": 4,
  "key": DummyCompositeKey {
    "id": 9,
    "name": "bbb",
  },
  "left": TreeNode {
    "height": 3,
    "key": DummyCompositeKey {
      "id": 5,
      "name": "fff",
    },
    "left": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 3,
        "name": "jjj",
      },
      "left": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 2,
          "name": "hhh",
        },
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 4,
          "name": "lll",
        },
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 1,
      "key": DummyCompositeKey {
        "id": 6,
        "name": "ggg",
      },
      "left": undefined,
      "right": undefined,
      "value": "g",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 3,
    "key": DummyCompositeKey {
      "id": 80,
      "name": "iii",
    },
    "left": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 70,
        "name": "mmm",
      },
      "left": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 10,
          "name": "aaa",
        },
        "left": undefined,
        "right": undefined,
        "value": "a",
      },
      "right": undefined,
      "value": "m",
    },
    "right": TreeNode {
      "height": 2,
      "key": DummyCompositeKey {
        "id": 90,
        "name": "ddd",
      },
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": DummyCompositeKey {
          "id": 190,
          "name": "eee",
        },
        "value": "e",
      },
      "value": "d",
    },
    "value": "i",
  },
  "value": "b",
}
`;

exports[`AVL Tree primitive keys add adds nested nodes 1`] = `
TreeNode {
  "height": 4,
  "key": 9,
  "left": TreeNode {
    "height": 3,
    "key": 5,
    "left": TreeNode {
      "height": 2,
      "key": 3,
      "left": TreeNode {
        "height": 1,
        "key": 2,
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": 4,
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 2,
      "key": 6,
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": 8,
        "left": undefined,
        "value": "c",
      },
      "value": "g",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 3,
    "key": 90,
    "left": TreeNode {
      "height": 2,
      "key": 70,
      "left": TreeNode {
        "height": 1,
        "key": 10,
        "left": undefined,
        "right": undefined,
        "value": "a",
      },
      "right": TreeNode {
        "height": 1,
        "key": 80,
        "left": undefined,
        "value": "i",
      },
      "value": "k",
    },
    "right": TreeNode {
      "height": 1,
      "key": 190,
      "value": "e",
    },
    "value": "d",
  },
  "value": "b",
}
`;

exports[`AVL Tree primitive keys add adds nested nodes 2`] = `
TreeNode {
  "height": 4,
  "key": 9,
  "left": TreeNode {
    "height": 3,
    "key": 5,
    "left": TreeNode {
      "height": 2,
      "key": 3,
      "left": TreeNode {
        "height": 1,
        "key": 2,
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": 4,
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 2,
      "key": 7,
      "left": TreeNode {
        "height": 1,
        "key": 6,
        "left": undefined,
        "right": undefined,
        "value": "g",
      },
      "right": TreeNode {
        "height": 1,
        "key": 8,
        "left": undefined,
        "value": "c",
      },
      "value": "m",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 3,
    "key": 90,
    "left": TreeNode {
      "height": 2,
      "key": 70,
      "left": TreeNode {
        "height": 1,
        "key": 10,
        "left": undefined,
        "right": undefined,
        "value": "a",
      },
      "right": TreeNode {
        "height": 1,
        "key": 80,
        "left": undefined,
        "value": "i",
      },
      "value": "k",
    },
    "right": TreeNode {
      "height": 1,
      "key": 190,
      "value": "e",
    },
    "value": "d",
  },
  "value": "b",
}
`;

exports[`AVL Tree primitive keys remove removes nested nodes 1`] = `
TreeNode {
  "height": 4,
  "key": 10,
  "left": TreeNode {
    "height": 3,
    "key": 5,
    "left": TreeNode {
      "height": 2,
      "key": 3,
      "left": TreeNode {
        "height": 1,
        "key": 2,
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": 4,
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 2,
      "key": 6,
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": 8,
        "left": undefined,
        "value": "c",
      },
      "value": "g",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 3,
    "key": 90,
    "left": TreeNode {
      "height": 2,
      "key": 70,
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": 80,
        "left": undefined,
        "value": "i",
      },
      "value": "k",
    },
    "right": TreeNode {
      "height": 1,
      "key": 190,
      "value": "e",
    },
    "value": "d",
  },
  "value": "a",
}
`;

exports[`AVL Tree primitive keys remove removes nested nodes 2`] = `
TreeNode {
  "height": 4,
  "key": 10,
  "left": TreeNode {
    "height": 3,
    "key": 5,
    "left": TreeNode {
      "height": 2,
      "key": 3,
      "left": TreeNode {
        "height": 1,
        "key": 2,
        "right": undefined,
        "value": "h",
      },
      "right": TreeNode {
        "height": 1,
        "key": 4,
        "value": "l",
      },
      "value": "j",
    },
    "right": TreeNode {
      "height": 2,
      "key": 6,
      "left": undefined,
      "right": TreeNode {
        "height": 1,
        "key": 8,
        "left": undefined,
        "value": "c",
      },
      "value": "g",
    },
    "value": "f",
  },
  "right": TreeNode {
    "height": 2,
    "key": 90,
    "left": TreeNode {
      "height": 1,
      "key": 70,
      "left": undefined,
      "right": undefined,
      "value": "k",
    },
    "right": TreeNode {
      "height": 1,
      "key": 190,
      "value": "e",
    },
    "value": "d",
  },
  "value": "a",
}
`;
